### PR TITLE
Firewall ipset GUI

### PR DESCRIPTION
**Pull request by systemcrash**
Needs some feedback before commit. 

Variously tested on 22.03.2. This is unlikely to land in 22.03.2, unless someone specially pulls/builds with this PR, so we can probably remove the loadfile GUI warning. See also [this](https://github.com/openwrt/openwrt/issues/11706) which seems related. 

We could probably drop the `if...else` blocks that check for fw4 if there are no cases where units would run e.g. fw3 in builds from now onward.

Edit: displays string for firewall version (4 or not 4).

ipsets_grid

ipsets_modal

### CODE REVIEW

**Comment by jow-** [comment]
Great works! Two remarks:

- The `src_` and `dest_` prefix for the type matches is optional and defaults to `src`. It can also be specified in the rule referencing the set (e.g. `list ipset 'setname src,dest'` to match the first field match of the set in the source direction and the second in the destination direction). This makes set declarations more flexible as they can be shared by both usptream and downstream direction rules. I suggest to drop the `src_` and `dest_` prefixes in the predefined choice list or at least add non-prefixed choices before the prefixed ones. Also add a human readable label for the predefined choices, e.g. `o.value('ip', _('Match IP address ("ip")'))` or `o.value('src_ip', _('Match source IP address ("src_ip")'))`.
- Please remove the "boot hang" warning. By the time the support will be merged in 22.03, it is long fixed. Also the issue is of a temporary nature while the ui description will stay for a very long time

**Comment by jow-** [comment]
> if the firewall bump does not make it into 22.03.3

The boot fix was added a few days after the 22.03.2 release already, it will definitely be part of 22.03.3

**Comment by systemcrash** [comment]
last push changed 'enabled' into checkboxes outside of modal, and put it to the last column on right, more in line to be consistent with other dialogues.

**Comment by systemcrash** [comment]
LGTM

**Review by feckert — COMMENTED** [review]
Just a small remark

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			o.depends('storage', 'list');
+			o.modalonly = true;
+
+
+			o = s.option(form.Value, 'hashsize', _('Initial Hash Size'));
+			o.depends('storage', 'hash');
+			o.placeholder = _('1024');
+			o.modalonly = true;
+
+		}
```
[the remark this answers is not in the document]
Comment (systemcrash) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js]: fixed

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+			o.depends({storage: 'hash', match: /_ip|_net|_mac/ });
+
+
+			o = s.option(form.Value, 'portrange', _('Port range'),
+				_('fromport-toport'));
+			o.datatype = 'neg(portrange)';
+			o.depends({family: 'ipv4', storage: 'bitmap', match: /_port/ });
+			o.depends({family: 'ipv4', storage: 'hash', match: /_port/ });
+			o.depends({family: 'ipv6', storage: 'hash', match: /_port/ });
+
```
[the remark this answers is not in the document]
Comment (systemcrash) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js]: fixed

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+		o.value('mac', _('mac: MAC addr'));
+		o.value('net', _('net: (sub)net'));
+		o.value('set', _('set: ipset*'));
+		o.value('src_ip', _('src_ip: Source IP'));
+		o.value('src_port', _('src_port: Source Port №'));
+		o.value('src_mac', _('src_mac: Source MAC addr'));
+		o.value('src_net', _('src_net: Source (sub)net'));
+		o.value('src_set', _('src_Set: Source ipset*')); // fw4 unsupported
+		o.value('dest_ip', _('dest_ip: Destination IP'));
+		o.value('dest_port', _('dest_port: Destination Port №'));
```
[the remark this answers is not in the document]
Comment (systemcrash) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:93]: https://www.google.com/search?&q=%E2%84%96

[the remark this answers is not in the document]
Comment (systemcrash) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:93]: Potentially superfluous, although "number" is frequently associated with "port".

**Comment by systemcrash** [comment]
Hi @jow- any final comments?

**Comment by jow-** [comment]
The `№` symbol looks weird in my font and is nowhere else used in the ui [1]. In the interest of consistency, please remove it and just spell it `Port`.

1:

```
$ grep -riE '\bport\b' modules/luci-base/po/templates/
modules/luci-base/po/templates/base.pot:msgid "Bridge port specific options"
modules/luci-base/po/templates/base.pot:msgid "DAE-Port"
modules/luci-base/po/templates/base.pot:msgid "DNS query port"
modules/luci-base/po/templates/base.pot:msgid "DNS server port"
modules/luci-base/po/templates/base.pot:msgid "Destination port"
modules/luci-base/po/templates/base.pot:msgid "Destination port"
modules/luci-base/po/templates/base.pot:msgctxt "VLAN port state"
modules/luci-base/po/templates/base.pot:msgid "Dynamic Authorization Extension port."
modules/luci-base/po/templates/base.pot:msgctxt "VLAN port state"
modules/luci-base/po/templates/base.pot:msgctxt "VLAN port state"
modules/luci-base/po/templates/base.pot:"Secure\">HTTPS</abbr> port."
modules/luci-base/po/templates/base.pot:msgid "Endpoint Port"
modules/luci-base/po/templates/base.pot:msgid "Expected port number."
modules/luci-base/po/templates/base.pot:msgid "External system log server port"
modules/luci-base/po/templates/base.pot:msgid "Fixed source port for outbound DNS queries."
modules/luci-base/po/templates/base.pot:msgid "Full port randomization"
modules/luci-base/po/templates/base.pot:msgid "Listen Port"
modules/luci-base/po/templates/base.pot:msgid "Listening port for inbound DNS queries."
modules/luci-base/po/templates/base.pot:msgid "Mirror monitor port"
modules/luci-base/po/templates/base.pot:msgid "Mirror source port"
modules/luci-base/po/templates/base.pot:"No fixed interface listening port defined, peers might not be able to "
modules/luci-base/po/templates/base.pot:"a non-standard Relay To port(<code>addr#port</code>)."
modules/luci-base/po/templates/base.pot:msgid "Optional. Port of peer."
modules/luci-base/po/templates/base.pot:msgid "Optional. UDP port used for outgoing and incoming packets."
modules/luci-base/po/templates/base.pot:msgid "Port"
modules/luci-base/po/templates/base.pot:msgctxt "WireGuard listen port"
modules/luci-base/po/templates/base.pot:msgid "Port %d"
modules/luci-base/po/templates/base.pot:msgid "Port isolation"
modules/luci-base/po/templates/base.pot:msgid "Port status:"
modules/luci-base/po/templates/base.pot:msgctxt "VLAN port state"
modules/luci-base/po/templates/base.pot:msgid "RADIUS Accounting Port"
modules/luci-base/po/templates/base.pot:msgid "RADIUS Authentication Port"
modules/luci-base/po/templates/base.pot:msgid "Randomize source port mapping"
modules/luci-base/po/templates/base.pot:msgctxt "nft redirect to port"
modules/luci-base/po/templates/base.pot:msgid "Redirect to local port <strong>%h</strong>"
modules/luci-base/po/templates/base.pot:msgctxt "nft dnat ip to addr:port"
modules/luci-base/po/templates/base.pot:msgid "Rewrite destination to <strong>%h</strong>, port <strong>%h</strong>"
modules/luci-base/po/templates/base.pot:msgctxt "nft dnat ip6 to addr:port"
modules/luci-base/po/templates/base.pot:msgid "Rewrite destination to <strong>%h</strong>, port <strong>%h</strong>"
modules/luci-base/po/templates/base.pot:msgctxt "nft snat ip to addr:port"
modules/luci-base/po/templates/base.pot:msgid "Rewrite source to <strong>%h</strong>, port <strong>%h</strong>"
modules/luci-base/po/templates/base.pot:msgctxt "nft snat ip6 to addr:port"
modules/luci-base/po/templates/base.pot:msgid "Rewrite source to <strong>%h</strong>, port <strong>%h</strong>"
modules/luci-base/po/templates/base.pot:msgid "SSH server port"
modules/luci-base/po/templates/base.pot:msgid "Source port"
modules/luci-base/po/templates/base.pot:msgid "Switch port"
modules/luci-base/po/templates/base.pot:msgid "TCP destination port"
modules/luci-base/po/templates/base.pot:msgid "TCP source port"
modules/luci-base/po/templates/base.pot:"segments. Often there is by default one Uplink port for a connection to the "
modules/luci-base/po/templates/base.pot:msgid "Transport header destination port"
modules/luci-base/po/templates/base.pot:msgid "Transport header source port"
modules/luci-base/po/templates/base.pot:msgid "UDP destination port"
modules/luci-base/po/templates/base.pot:msgid "UDP source port"
modules/luci-base/po/templates/base.pot:msgid "VEPA (Virtual Ethernet Port Aggregator)"
modules/luci-base/po/templates/base.pot:msgid "VPN Local port"
modules/luci-base/po/templates/base.pot:msgid "VPN Server port"
modules/luci-base/po/templates/base.pot:msgid "valid IPv4 address:port"
modules/luci-base/po/templates/base.pot:msgid "valid address:port"
modules/luci-base/po/templates/base.pot:msgid "valid host:port"
modules/luci-base/po/templates/base.pot:msgid "valid port or port range (port1-port2)"
modules/luci-base/po/templates/base.pot:msgid "valid port value"
```

**Comment by systemcrash** [comment]
Done

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/snats.js [hunk 4]
```diff
@@ -235,6 +235,10 @@ return view.extend({
 		o.depends({ proto: 'tcp', '!contains': true });
 		o.depends({ proto: 'udp', '!contains': true });
 
+		o = s.taboption('advanced', form.Value, 'ipset', _('Use ipset'), _('Unsupported in fw4'));
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/snats.js:238]: Please wrap this option declaration in `if (!L.hasSystemFeature('firewall4')) { ... }` and remove the `Unsupported in fw4` remark.

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+return view.extend({
+
+	load: function() {
+		return Promise.all([
+			uci.load('firewall')
+		]);
+	},
+
+	render: function(data) {
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:17]: Drop the SNAT migration code from this view.

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	renderIPsets: function(data) {
+		var boardinfo = data[0],
+		    m, s, o;
+
+		m = new form.Map('firewall', _('Firewall - IP sets'),
+			_('firewall4 supports referencing and creating IP sets to simplify matching of large address lists without the need to create one rule per item to match. Port ranges in ipsets are unsupported by firewall4.<br />' + 
+				'<a href="https://ipset.netfilter.org/ipset.man.html">Reference</a>.'));
+
+
+		if (L.hasSystemFeature('firewall4')) {
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:33]: This does not belong here imho, if at all it should appear on the zone overview page.

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+			_('Packet fields to match upon.<br />' +
+			  'Syntax: <em>direction_datatype</em>. e.g.: <code>src_port, dest_net</code>.<br />' +
+			  'Directions: <code>src, dst</code>. Datatypes: <code>ip, port, mac, net, set</code>.<br />' +
+			  'Direction prefixes are optional.<br />' +
+			  '*Note: datatype <code>set</code> is unsupported in fw4.'));
+		o.value('ip', _('ip: IP addr'));
+		o.value('port', _('port: Port'));
+		o.value('mac', _('mac: MAC addr'));
+		o.value('net', _('net: (sub)net'));
+		o.value('set', _('set: ipset*'));
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:86]: Guard the unsupported `set` choice by `if (!L.hasSystemFeature('firewall4'))`  (consider declaring a global helper bool `var fw4 = L.hasSystemFeature('firewall4')` to allow simplifying this and other checks)

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+		o.value('src_ip', _('src_ip: Source IP'));
+		o.value('src_port', _('src_port: Source Port'));
+		o.value('src_mac', _('src_mac: Source MAC addr'));
+		o.value('src_net', _('src_net: Source (sub)net'));
+		o.value('src_set', _('src_Set: Source ipset*')); // fw4 unsupported
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:91]: Guard the unsupported `set` choice by `if (!L.hasSystemFeature('firewall4'))`

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+		o.value('dest_ip', _('dest_ip: Destination IP'));
+		o.value('dest_port', _('dest_port: Destination Port'));
+		o.value('dest_mac', _('dest_mac: Destination MAC addr'));
+		o.value('dest_net', _('dest_net: Destination (sub)net'));
+		o.value('dest_set', _('dest_set: Destination ipset*')); // fw4 unsupported
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:96]: Guard the unsupported `set` choice by `if (!L.hasSystemFeature('firewall4'))`

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/forwards.js [hunk 10]
```diff
@@ -194,6 +194,10 @@ return view.extend({
 		o.nocreate = true;
 		o.default = 'wan';
 
+		o = s.taboption('advanced', form.Value, 'ipset', _('Use ipset'));
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/forwards.js:197]: We should loop uci sections of type `ipset` and present their names here, like

```javascript
o = s.taboption('advanced', form.Value, 'ipset', _('Use ipset'));
uci.sections('firewall', 'ipset', function(s) {
    if (typeof(s.name) == 'string')
            this.value(s.name, s.comment ? '%s (%s)'.format(s.name, s.comment) : s.name);
});
```

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/rules.js [hunk 11]
```diff
@@ -349,6 +349,10 @@ return view.extend({
 		o.allowany = true;
 		o.allowlocal = 'src';
 
+		o = s.taboption('general', form.Value, 'ipset', _('Use ipset'));
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/rules.js:352]: We should loop uci sections of type `ipset` and present their names here, like

```javascript
o = s.taboption('advanced', form.Value, 'ipset', _('Use ipset'));
uci.sections('firewall', 'ipset', function(s) {
    if (typeof(s.name) == 'string')
            this.value(s.name, s.comment ? '%s (%s)'.format(s.name, s.comment) : s.name);
});
```

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/snats.js (hunk 4 above)
Comment (jow-) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/snats.js:238]: We should loop uci sections of type `ipset` and present their names here, like

```javascript
o = s.taboption('advanced', form.Value, 'ipset', _('Use ipset'));
uci.sections('firewall', 'ipset', function(s) {
    if (typeof(s.name) == 'string')
            this.value(s.name, s.comment ? '%s (%s)'.format(s.name, s.comment) : s.name);
});
```

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/rules.js (hunk 11 above)
Comment (systemcrash) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/rules.js:352]: Nice addition. Borrowed.

**Comment by systemcrash** [comment]
Resolved.

File: applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+			_('Packet fields to match upon.<br />' +
+			  'Syntax: <em>direction_datatype</em>. e.g.: <code>src_port, dest_net</code>.<br />' +
+			  'Directions: <code>src, dst</code>. Datatypes: <code>ip, port, mac, net, set</code>.<br />' +
+			  'Direction prefixes are optional.<br />' +
+			  '*Note: datatype <code>set</code> is unsupported in fw4.'));
+		o.value('ip', _('ip: IP addr'));
+		o.value('port', _('port: Port'));
+		o.value('mac', _('mac: MAC addr'));
+		o.value('net', _('net: (sub)net'));
+		if (have_fw4)
```
Comment (jow-) [inline applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js:78]: You need to invert this and the other two, rest lgtm

Comment (systemcrash) [reply applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/ipsets.js]: XD well spotted 😄

**Comment by systemcrash** [comment]
Resolved.

**Comment by systemcrash** [comment]
@jow- apparently you already picked these into master. I guess I can close this? I was confused as to why my rebase was coming up empty 😉

**Pull request closed without merging**